src/codegeneration/ForOnTransformer.js

```diff
@@ -0,0 +1,70 @@
+// Copyright 2015 Traceur Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {VARIABLE_DECLARATION_LIST} from '../syntax/trees/ParseTreeType.js';
+import {TempVarTransformer} from './TempVarTransformer.js';
+import {
+  createIdentifierExpression as id,
+  createMemberExpression,
+  createVariableStatement
+} from './ParseTreeFactory.js';
+import {
+  parseStatement,
+  parseExpression
+} from './PlaceholderParser.js';
+
+/**
+ * Desugars for-on statement.
+ */
+export class ForOnTransformer extends TempVarTransformer {
+  /**
+   * @param {ForOnStatement} original
+   * @return {ParseTree}
+   */
+  transformForOnStatement(original) {
+    var tree = super.transformForOnStatement(original);
+    var observer = id(this.getTempIdentifier());
+    var result = id(this.getTempIdentifier());
+    var resolve = id(this.getTempIdentifier());
+    var reject = id(this.getTempIdentifier());
+
+    var assignment;
+    if (tree.initializer.type === VARIABLE_DECLARATION_LIST) {
+      // {var,let} initializer = $result.value;
```

Comment:
In ES6 the loop initializer, if being a variable declaration may not have an initializer RHS.

```js
for (var x = 1 in {}) {}  // SyntaxError
for (var x = 1 of []) {}  // SyntaxError
```